src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -869,6 +1183,59 @@ public void HandleEvent(SocketAsyncContext context)
                 op.Dispatch();
             }
 
+            internal void BatchOrDispatch(SocketAsyncContext context, Span<Interop.Sys.IoControlBlock> ioControlBlocks, Span<AsyncOperation> batchedOperations, ref int batchedCount)
+            {
+                AsyncOperation nextOperation;
+                using (Lock())
+                {
+                    switch (_state)
+                    {
+                        case QueueState.Ready:
+                            Debug.Assert(_tail == null, "State == Ready but queue is not empty!");
+                            _sequenceNumber++;
+                            return;
+
+                        case QueueState.Waiting:
+                            Debug.Assert(_tail != null, "State == Waiting but queue is empty!");
+                            _state = QueueState.Processing;
+                            nextOperation = _tail.Next;
+
+                            while (batchedCount < ioControlBlocks.Length && nextOperation.TryBatch(context, batchedCount, ref ioControlBlocks[batchedCount]))
+                            {
+                                batchedOperations[batchedCount++] = nextOperation;
+                                _batchedOperationsCount++;
+
+                                if (nextOperation == nextOperation.Next)
+                                {
+                                    // we have batched some operations and reached the end of the queue
```

Comment:
this is weird: the last operation in the queue points to itself via the `Next` property

Reply:
fyi, the last operation points to the first operation.
When the operation points to itself, the queue has a single element.